### PR TITLE
155969902 retrieve fail management

### PR DESCRIPTION
**Pull request by r3vit**

### CODE REVIEW

File: httpclient/httpclient.go [hunk 1]
```diff
@@ -29,16 +30,23 @@ func GetURL(URL string) ([]byte, RespStatus, error) {
...
 			return nil, RespStatus{Status: err.Error(), StatusCode: -1}, err
 		}
+
+		// Set headers.
+		for k, v := range headers {
+			req.Header.Add(k, v)
+		}
+
 		// Set special user agent for bot.
-		req.Header.Set("User-Agent", "Golang_talia_backend_bot/0.0.1")
```
Comment (lussoluca) [inline httpclient/httpclient.go]: s/Golang_talia_backend_bot/Golang_italia_backend_bot

File: httpclient/httpclient.go [hunk 2]
```diff
@@ -29,16 +30,23 @@ func GetURL(URL string) ([]byte, RespStatus, error) {
...
+		req.Header.Add("User-Agent", "Golang_talia_backend_bot/0.0.1")
```
Comment (lussoluca) [inline httpclient/httpclient.go]: Can we avoid the hardcoded version here?

File: crawler/bitbucket.go [hunk 3]
```diff
@@ -32,29 +37,81 @@ type response struct {
 
 // GetRepositories retrieves the list of all repository from an hosting.
 func (host Bitbucket) GetRepositories(repositories chan Repository) error {
-	var nextPage = host.URL
+	var sourceURL = host.URL
+
+	// Redis connection.
+	redisClient, err := redisClientFactory("redis:6379")
```
Comment (lussoluca) [inline crawler/bitbucket.go]: we should invert the logic here, the Bitbucket implementation doesn't have to knows nothing about the fail management.

File: cmd/all.go [hunk 4]
```diff
@@ -135,3 +111,29 @@ func splitFullName(fullName string) (string, string) {
 	s := strings.Split(fullName, "/")
 	return s[0], s[1]
 }
+
+func initPrometheus() prometheus.Counter {
```
Comment (lussoluca) [inline cmd/all.go]: maybe we could move the Prometheus funcs in a prometheus package

File: crawler/bitbucket.go [hunk 5]
```diff
@@ -32,29 +37,81 @@ type response struct {
...
+		if err != nil {
+			return err
+		}
 
+		// Add repositories to the channel that will perform the check on everyone.
 		for _, v := range result.Values {
 			repositories <- Repository{
 				Name: v.FullName,
 				//URL:  v.Links.Clone[0].Href + "/raw/default/publiccode.yml",
-				URL: v.Links.Clone[0].Href + "/raw/default/.gitignore",
```
Comment (lussoluca) [inline crawler/bitbucket.go]: the name of the file to search must be loaded from an environment variable

File: crawler/hosting.go [hunk 6]
```diff
@@ -14,32 +14,66 @@ type Hosting struct {
...
+	}
+
+	// Manage every host
 	for i, hosting := range hostings {
 		switch hosting.ServiceName {
 		case "bitbucket":
-			hostings[i].ServiceInstance = Bitbucket{
+			// Check if there is an URL that wasn't correctly retrieved.
+			// URL.value="false" => set hosting.URL to the one that one ("false")
+			keys, _ := redisClient.Keys("*").Result()
```
Comment (lussoluca) [inline crawler/hosting.go]: how can I understand which hosting has been interrupted? Maybe we can use an hashmap to store the hosting name

File: cmd/all.go (hunk 4 above)
Comment (r3vit) [reply cmd/all.go]: I'll add a "metrics" package. It uses PrometheusCounter() as prometheus.Counter generator.

File: crawler/bitbucket.go (hunk 3 above)
Comment (r3vit) [reply crawler/bitbucket.go]: Ok. I moved redis logic outside GetRepositories(). Now it saves the sourceURL if and only if it failed to retrieve the repositories list.

File: crawler/crawler.go [hunk 7]
```diff
@@ -14,9 +19,24 @@ func Process(hosting Hosting, repositories chan Repository) {
 		return
 	}
 
-	err := hosting.ServiceInstance.GetRepositories(repositories)
+	sourceURL, err := hosting.ServiceInstance.GetRepositories(repositories)
 	if err != nil {
+		// Redis connection.
+		redisClient, err := redisClientFactory("redis:6379")
```
Comment (lussoluca) [inline crawler/crawler.go]: Can we inject this as func argument? Also we should read the host (`redis:6379`) from env

File: cmd/all.go [hunk 8]
```diff
@@ -82,7 +82,7 @@ func checkAvailability(fullName, url string, headers map[string]string, ch chan<
 	if status.StatusCode == http.StatusOK && err == nil {
 		// Save the file.
 		vendor, repo := splitFullName(fullName)
-		fileName := "gitignore"
+		fileName := os.Getenv("CRAWLED_FILENAME")
 		saveFile(vendor, repo, fileName, body)
 
 		ch <- fmt.Sprintf("%s - hit - %s", fullName, url)
```
Comment (lussoluca) [inline cmd/all.go]: What is the purpose of the `ch` channel?

Comment (r3vit) [reply cmd/all.go]: It's unused, I'll remove it.

**Pull request merged**